tests/defer/src/commonTest/kotlin/test/DeferNormalizedCacheTest.kt

```diff
@@ -0,0 +1,462 @@
+import com.apollographql.apollo3.api.ApolloResponse
+import com.apollographql.apollo3.api.Error
+import com.apollographql.apollo3.api.Operation
+import com.apollographql.apollo3.cache.normalized.ApolloStore
+import com.apollographql.apollo3.cache.normalized.FetchPolicy
+import com.apollographql.apollo3.cache.normalized.api.MemoryCacheFactory
+import com.apollographql.apollo3.cache.normalized.fetchPolicy
+import com.apollographql.apollo3.cache.normalized.optimisticUpdates
+import com.apollographql.apollo3.cache.normalized.store
+import com.apollographql.apollo3.exception.ApolloCompositeException
+import com.apollographql.apollo3.exception.ApolloException
+import com.apollographql.apollo3.exception.ApolloHttpException
+import com.apollographql.apollo3.exception.ApolloNetworkException
+import com.apollographql.apollo3.exception.CacheMissException
+import com.apollographql.apollo3.mockserver.MockResponse
+import com.apollographql.apollo3.mockserver.MockServer
+import com.apollographql.apollo3.mockserver.enqueueMultipart
+import com.apollographql.apollo3.network.NetworkTransport
+import com.apollographql.apollo3.testing.runTest
+import com.benasher44.uuid.uuid4
+import defer.WithFragmentSpreadsMutation
+import defer.WithFragmentSpreadsQuery
+import defer.fragment.ComputerFields
+import defer.fragment.ScreenFields
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.catch
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.toList
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertFails
+import kotlin.test.assertFailsWith
+import kotlin.test.assertIs
+
+@OptIn(ApolloExperimental::class)
+class DeferNormalizedCacheTest {
+  private lateinit var mockServer: MockServer
+  private lateinit var apolloClient: ApolloClient
+  private lateinit var store: ApolloStore
+
+  private suspend fun setUp() {
+    store = ApolloStore(MemoryCacheFactory())
+    mockServer = MockServer()
+    apolloClient = ApolloClient.Builder().serverUrl(mockServer.url()).store(store).build()
+  }
+
+  private suspend fun tearDown() {
+    mockServer.stop()
+    apolloClient.dispose()
+  }
+
+  @Test
+  fun cacheOnly() = runTest(before = { setUp() }, after = { tearDown() }) {
+    apolloClient = apolloClient.newBuilder().fetchPolicy(FetchPolicy.CacheOnly).build()
+
+    // Cache is empty
+    assertFailsWith<CacheMissException> {
+      apolloClient.query(WithFragmentSpreadsQuery()).execute()
+    }
+
+    // Fill the cache by doing a network only request
+    val jsonList = listOf(
+        """{"data":{"computers":[{"__typename":"Computer","id":"Computer1"}]},"hasNext":true}""",
+        """{"data":{"cpu":"386","year":1993,"screen":{"__typename":"Screen","resolution":"640x480"}},"path":["computers",0],"hasNext":true}""",
+        """{"data":{"isColor":false},"path":["computers",0,"screen"],"hasNext":false,"label":"a"}""",
+    )
```

Comment:
(Dog) Food for thought: looks like we could use test builders to simplify those tests